### PR TITLE
[Transform] Add a new transform op that applies patterns

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This revision introduces a `transform.iree.apply_patterns` operation that operates on
an isolated from above op and applies a set of patterns while listening and updating
transform dialect handles.

The list of patterns is specified via attributes that are additive and roughly play the role of
populate functions.

iree-opt is extended so that it can run transform dialect directly without files, which
results in general simplifications.

### CODE REVIEW

**Review by ThomasRaoux — APPROVED** [review]
LGTM

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/StructuredTransformOpsExt.h [hunk 1]
```diff
@@ -19,6 +20,52 @@ class LinalgOp;
 namespace scf {
 class ForOp;
 } // namespace scf
+
+class TrackingListener : public RewriteListener,
+                         public transform::TransformState::Extension {
+public:
+  // MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TrackingListener);
```
Comment (ThomasRaoux) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/StructuredTransformOpsExt.h:27]: nit: remove?

File: tools/iree-opt-main.cc [hunk 2]
```diff
@@ -24,6 +24,16 @@ int main(int argc, char **argv) {
...
   mlir::iree_compiler::registerHALTargetBackends();
 
+  // Also register the transform interpreter pass so that iree-opt can run
+  // transform dialect IR without resorting to a separate file.
+  // Resorting to separate files is a convenience for iree-compile to be able to
+  // use the transform dialect without requiring special plumbing.
+  // Still the preferred mode of execution should be to transport the relevant
+  // piece of transform IR in the right location, for each piece of code we
+  // want to transform for.
+  mlir::linalg::transform::registerTransformDialectInterpreterPass();
```
Comment (GMNGeoffrey) [inline tools/iree-opt-main.cc:34]: My understanding is that the transform dialect is supposed to remain optional (documented in our current OKRs). How is this in keeping with that goal?

Comment (MaheshRavishankar) [reply tools/iree-opt-main.cc:34]: But the dialect is in core. I think this is OK.

Comment (nicolasvasilache) [reply tools/iree-opt-main.cc:34]: Registration does not imply that it is mandatory to use it in production.
Optionality is builtin the whole process and even the whole MLIR codegen, pick and chose the best tool at hand for the job.

**Comment by nicolasvasilache** [comment]
> So does this mean all the other places where MLIR files are used as input will be dropped?

Not really, the problem of shipping the transform IR next to the module that will require a transformation recipe is still an intrusive process that needs to be designed properly: the IREE compiler doesn't yet know about it and drops that part of the IR; for using the transform IR in those places we still need a separate transform file (iree-compile, iree-mlir-run and friends). 

However using a file and wrapping IR in `hal.variant` is overkill for tests that can be run completely with iree-opt.
Those tests can indeed drop their file and embed the transform IR in the test directly.

**Pull request merged**